plugin/dnstap/handler.go

```diff
@@ -16,11 +16,14 @@ import (
 // Dnstap is the dnstap handler.
 type Dnstap struct {
 	Next plugin.Handler
-	Out  io.Writer
+	IO   IOThread
 	Pack bool
 }
 
```

Comment:
Ok, not in this PR but looking this over I am wondering why we didn't just put the Dnstap *in the context* rather than *extending the Context type*. Here: https://github.com/coredns/coredns/pull/1083/files#diff-9bba1e13b803e0b0fabe763bea4b5fcfR34

What @huynhmb359 and @rdrozhdzh are trying to do are two things:

1) Send a custom DNS tap message in a downstream middleware (see https://github.com/dnstap/golang-dnstap/pull/14).
2) Append some data to the CLIENT_RESPONSE message in a downstream middleware.

For the second one, the idea is that the downstream policy middleware will attach a bunch of EDNS0 options from the original request to the DNS Tap response message so we can correlate these things. So, what I was hoping we could do is have the Dnstap middleware stuff itself in the context, and also stuff a generic map[string]byte[] in the context with a specific name. Downstream middleware could then populate that extra data map by their name. When the DNStap middleware gets back control, it would see the data in that context and included it in the extra field of the CR message.

So...that's what I would like to see. I don't think it means we need to stuff the Dnstap into the Context right now, but I am not sure why we did it via the type.

And yes, I fully acknowledge that I should have said this a month or two ago!!! Clearly not the most thorough review I have done 😜 .

Reply:
EDNS0 options aren't into the response message?
Also the dnstap CLI tool cannot decode/encode 3rd party payloads.

Reply:
They would be in the request but aren't sent back to the client, so wouldn't be in the response.